### PR TITLE
Use standard M600 when mmu not detected

### PR DESCRIPTION
**Pull request by GMagician**
This is a first part change. I tested it without mmu (mine is not ready yet) but I think there are no issues when it's installed.
Purpose is to let printer works when mmu is temporary disabled.
In such case M600 works as standard and a filament change is done as usual.

If I'm not wrong this is what original Prusa firmware does.

Next change is to remove filament runout detection when mmu is working because I think finda should detect filament runout and not extruder sensor (it must be used only to properly load/unload filament up to/out of the extruder gears).

I need some "mmu2" user feedback about this beforte proceeeding.
Thanks

### CODE REVIEW

**Comment by thisiskeithb** [comment]
~~I've pinged some MMU2 owners to look at this.~~

Update: They say your comments are correct, but can't test the PR for one reason or another. 🙂

**Comment by JohnnyTheOne** [comment]
How can you temporarily disable mmu?

**Comment by GMagician** [comment]
@JohnnyTheOne simply not power it on or disconnect serial

**Comment by JohnnyTheOne** [comment]
Have tested this pr and it works with and without mmu

**Comment by GMagician** [comment]
@JohnnyTheOne thanks for test you did.
I'm wondering, since you have the H/W, what happens when sensor on extruders detect a missing filament.
It should execute a M600 but in my opinion it shouldn't. Only finda sensor should raise a filament missing script (mmu has it's own script fort filament runout), extruder sensor should only be used to properly load to gear/unload from gear the filament and then "muted".
What do you think?
Thanks

**Comment by JohnnyTheOne** [comment]
You are welcome, I am not an expert with mmu but if you disable the mmu finda will not work anymore. why not both when needed?

**Comment by GMagician** [comment]
I'm not going to disable finda but extruder runout sensor. I would like to use it just to "load to gear"/"unload from gear"

**Comment by GMagician** [comment]
@thisiskeithb I missed this:

> Update: They say your comments are correct, but can't test the PR for one reason or another. 🙂

Was that related to disable extruder sensor from "runout" detection and use it only for load/unload?

**Comment by JohnnyTheOne** [comment]
I understand what will happen if you turn off mmu temporarily? will extruder sensor keep working?

**Comment by GMagician** [comment]
> I understand what will happen if you turn off mmu temporarily? will extruder sensor keep working?

When mmu is "disabled" extruder sensor has to work as usual

**Comment by GMagician** [comment]
@JohnnyTheOne if you have some time to test new version "extruder sensor disabled when working with mmu2" this will speed up things because my H/W is not ready yet.
Thanks in advance

**Comment by JohnnyTheOne** [comment]
Now get this error 
```
In file included from Marlin\src\feature/runout.h:41,
                 from Marlin\src\MarlinCore.cpp:190:
Marlin\src\feature/../feature/mmu/mmu2.h: In static member function 'static void MMU2::set_runout_valid(bool)':
Marlin\src\feature/../feature/mmu/mmu2.h:105:18: error: 'runout' was not declared in this scope
  105 |       if (valid) runout.reset();
      |                  ^~~~~~
*** [.pio\build\BIGTREE_GTR_V1_0\src\src\MarlinCore.cpp.o] Error 1
```

**Comment by GMagician** [comment]
Thanks @JohnnyTheOne, I have fixed it. I think issue was due to nested include calls

**Comment by JohnnyTheOne** [comment]
Have just tested again runout script is not triggered keeps printing. continue testing tomorrow

**Comment by GMagician** [comment]
> 
> 
> Have just tested again runout script is not triggered keeps printing. continue testing tomorrow

Yep, only finda should trigger runout. That's the idea. Thanks again

**Comment by JohnnyTheOne** [comment]
Have also tested finda not working either

**Comment by GMagician** [comment]
That sound weird. mmu2 runout is indipendent. It get finda status from P0 command. You may try to enable MMU2_DEBUG and check if pinda reports its status correctly. you should read in console something like 

> MMU => 1/0
> MMU - ENABLED

**Comment by JohnnyTheOne** [comment]
Ok i tested again today. if i remove filament from finda during printing, printer restarts I lose connection to octoprint

**Comment by GMagician** [comment]
That's weird than the other. I can't understand why such change should lead to what you see.
I think that since you are usign octoprint is not possible to see terminal log and check what happens.
I think I have to park this until I'm ready with my HW, but I suspect it will be a looooong wait

**Comment by JohnnyTheOne** [comment]
one more time with repeter host [mmu test log.txt](https://github.com/MarlinFirmware/Marlin/files/6554364/mmu.test.log.txt)
with octoprint printer restart and repetier host finda not working? I am now going to test LCD via sd.

ok i now understand a little more of what is happening. if filamet in finda is present at startup runout script will not be triggered. if i first give T # command and then remove filament then printer restarts

**Comment by GMagician** [comment]
Ok, from your log I see:

> 15:19:04.343 : MMU => 1
> 15:19:04.343 : MMU - ENABLED

then you started with finda=1 (filament present)

but I can't see any other MMU message.
Every 300ms Marlin should send a P0 command and handle finda status and you may check it by removing commented code in mmu2.cpp
```c++
// This is super annoying. Only activate if necessary
// if (finda_runout_valid) DEBUG_ECHOLNPAIR_F("MMU <= 'P0'\nMMU => ", finda, 6);
```
lines 294-295

What I see is that when executing C0 and filament is missing something like:
```c++
      else {
        while (planner.movesplanned() < 3) {
          current_position.e += 0.25;
          line_to_current_position(MMM_TO_MMS(120));
        }
```

is done. Not sure if a fast board can empty planner before it add next command (hence doing an infinite loop)

**Comment by GMagician** [comment]
But does original code works in your configuration? Or you still have issues even there?

**Comment by JohnnyTheOne** [comment]
> But does original code works in your configuration? Or you still have issues even there?

original code partially works. extruder sensor work; runout start; and if i remove filament from finda, printer restart

**Comment by thinkyhead** [comment]
It might help to enable the `POSTMORTEM_DEBUGGING` to see why the firmware is rebooting. It might only be the watchdog getting triggered, but could be something else.

**Comment by thinkyhead** [comment]
I don't know if this is related to the crashing behavior, but I found and fixed one small glitch:
```diff
- resume_print(0, 0, 0, beep_count DXC_PASS);
+ resume_print(0, 0, 0, beep_count, 0 DXC_PASS);
```

**Comment by thinkyhead** [comment]
I've merged the latest code from bugfix, so please give this a re-test when you have a chance to see if any bugs remain.

I don't know if this is related to the crashing behavior, but I found and fixed one small glitch:
```diff
- resume_print(0, 0, 0, beep_count DXC_PASS);
+ resume_print(0, 0, 0, beep_count, 0 DXC_PASS);
```

File: Marlin/src/MarlinCore.cpp [hunk 1]
```diff
@@ -811,7 +811,10 @@ void idle(bool no_stepper_sleep/*=false*/) {
   (void)check_tool_sensor_stats(active_extruder, true);
 
   // Handle filament runout sensors
-  TERN_(HAS_FILAMENT_SENSOR, runout.run());
+  #if HAS_FILAMENT_SENSOR
+    if (TERN0(HAS_PRUSA_MMU2, !mmu2.enabled()))
```
Comment (GMagician) [inline Marlin/src/MarlinCore.cpp]: Not sure to understand correctly these macros but shouldn't this be TERN1? runout has to be called when HAS_PRUSA_MMU2 is not defined

Comment (thinkyhead) [reply Marlin/src/MarlinCore.cpp]: Ha, yes! Leave it to me, right?

Comment (GMagician) [reply Marlin/src/MarlinCore.cpp]: I was not sure...

**Comment by GMagician** [comment]
@JohnnyTheOne are you so kind to test new fixes Thinkyhead did? My mmu is not ready and I think it will never be ready if things are going this way...

Edit: not related to this PR of course

**Comment by JohnnyTheOne** [comment]
>  on 31 May
> with octoprint printer restart and repetier host finda not working?
> 

With repetier host printer restart and octoprint finda not working is the opposite?

**Comment by thinkyhead** [comment]
Maybe the crash is occurring at this point:
```cpp
   #if HAS_FILAMENT_SENSOR
     if (TERN1(HAS_PRUSA_MMU2, !mmu2.enabled()))
       runout.run();
   #endif
```
Try commenting out this section to see if it still crashes.

**Comment by JohnnyTheOne** [comment]
In which file?

**Comment by thisiskeithb** [comment]
> In which file?

`MarlinCore.cpp`: https://github.com/MarlinFirmware/Marlin/blob/759d04cdd0a1b04a2acd780462d0e47107bde075/Marlin/src/MarlinCore.cpp#L813-L817

**Comment by JohnnyTheOne** [comment]
Commenting out section no longer crashes but finda no longer works

**Comment by GMagician** [comment]
Maybe I miss something but:
1) why should this happens? runout.run is not called when mmu2 is enabled, hence commenting it out or not shouldn't change anything.
2) Finda is handled somewhere else (read by serial) and should use its own gcode define (same `M600`)

really weird

**Comment by thinkyhead** [comment]
> commenting it out or not shouldn't change anything

And yet, I suggested this because I was pretty sure it would have the reported effect. 🤷🏻

Since `mmu2.enabled()` is almost certainly returning `false` the crash is probably somewhere inside of `runout.run()`.

So next, try this:
```cpp
   if (TERN1(HAS_PRUSA_MMU2, !mmu2.enabled()))
     SERIAL_ECHOLNPGM("wants to call runout.run()");
```
We might also want to look at serial port issues and pin conflicts. Some SKR boards have been having serial issues lately, which might affect other STM32-based boards as well.

Let's enable `PINS_DEBUGGING` and then have a look at the output of `M43` from your printer to rule out pin issues.

**Comment by JohnnyTheOne** [comment]
It still crashes
```cpp
  // Handle filament runout sensors
  #if HAS_FILAMENT_SENSOR
   if (TERN1(HAS_PRUSA_MMU2, !mmu2.enabled()))
     SERIAL_ECHOLNPGM("wants to call runout.run()");
  #endif
```
I use serial 6 wifi port and for M43 see log file [Log #21865.txt](https://github.com/MarlinFirmware/Marlin/files/6912271/Log.21865.txt)

**Comment by GMagician** [comment]
> Since mmu2.enabled() is almost certainly returning false the crash is probably somewhere inside of runout.run().

When mmu2 is connected it should return true. Enabling debug status is shown. Looking to an old @JohnnyTheOne post, mmu2 is enabled:

> 15:19:04.343 : MMU => 1
> 15:19:04.343 : MMU - ENABLED

**Comment by thinkyhead** [comment]
> It still crashes

Interesting. That seems to imply that the simple act of calling `mmu2.enabled()` is enough to cause a crash. So let's try just skipping that…
```cpp
  // Handle filament runout sensors
  #if HAS_FILAMENT_SENSOR
   //if (TERN1(HAS_PRUSA_MMU2, !mmu2.enabled()))
     runout.run();
  #endif
```

> When mmu2 is connected it should return true.

- The value of `_enabled` is first set to `true` by `mmu2.mmu_loop()` when `state` gets to `-4`.
- The `idle()` function has a call to `mmu2.mmu_loop()` near the end. Could the crash be occurring there?
- The `mmu2.state` will only get to `-4` after `mmu2.init()` and a few calls to `mmu2.mmu_loop()`.
- The `runout` code above is also in `idle()` and is called _before_ `mmu2.mmu_loop()`. Maybe after would be better, but it should not make a difference.

> …see log file…

In the log file I only see these MMU messages…
```
MMU <= T0
MMU => 'ok'
MMU <= 'C0'
MMU => 'ok'
MMU can_load:
```
…but I don't see `MMU => 'start'` or `MMU - ENABLED` or any other output to indicate that `mmu2.init` got called. Maybe the log is simply incomplete?

**Comment by thinkyhead** [comment]
The `M43` output seems fine.

**Comment by JohnnyTheOne** [comment]
```cpp
  // Handle filament runout sensors
  #if HAS_FILAMENT_SENSOR
   //if (TERN1(HAS_PRUSA_MMU2, !mmu2.enabled()))
     runout.run();
  #endif
```
Finda doesn't work but extruder does [Log #21865_1.txt](https://github.com/MarlinFirmware/Marlin/files/6912451/Log.21865_1.txt)

**EDIT: Three backticks ` ``` ` for mult-line code.**

**Comment by GMagician** [comment]
> …but I don't see MMU => 'start' or MMU - ENABLED or any other output to indicate that mmu2.init got called. Maybe the log is simply incomplete?

New log is correctly reporting mmu is enabled

**Pull request merged**